clients/wallet/src/stellar_wallet.rs

```diff
@@ -374,12 +378,113 @@ impl StellarWallet {
 			operations,
 		)?;
 
-		let _ = self.cache.save_tx_envelope(envelope.clone())?;
-
 		self.submit_transaction(envelope).await
 	}
 }
 
+pub enum Tx {
+	V0(TransactionV0),
+	V1(Transaction),
+}
+
+// Error handling methods for StellarWallet.
+impl StellarWallet {
+	pub async fn handle_error(&self, error: Error) -> Result<TransactionResponse, Error> {
+		match &error {
+			Error::HorizonSubmissionError { title: _, status: _, reason, envelope_xdr } => {
+				let transaction_envelope;
+				if let Some(env_xdr) = envelope_xdr {
+					transaction_envelope =
+						TransactionEnvelope::from_base64_xdr(env_xdr).expect("should convert ok");
+					match &reason[..] {
+						"tx_bad_seq" =>
+							self.bump_sequence_number_and_submit(transaction_envelope).await,
+						"tx_internal_error" => self.submit_transaction(transaction_envelope).await,
+						_ => Err(error),
+					}
+				} else {
+					Err(error)
+				}
+			},
+
+			Error::CacheError(CacheError {
+				kind: CacheErrorKind::SequenceNumberAlreadyUsed,
+				path: _,
+				envelope,
+				sequence_number,
+			}) => {
+				if let Some(transaction_envelope) = envelope {
+					return self.bump_sequence_number_and_submit(transaction_envelope.clone()).await
+				}
+				Err(error)
+			},
+			_ => Err(error),
+		}
+	}
+
+	fn get_transaction_from_envelope(&self, tx_envelope: TransactionEnvelope) -> Option<Tx> {
+		match tx_envelope {
+			TransactionEnvelope::EnvelopeTypeTxV0(transaction) => Some(Tx::V0(transaction.tx)),
+			TransactionEnvelope::EnvelopeTypeTx(transaction) => Some(Tx::V1(transaction.tx)),
+			TransactionEnvelope::EnvelopeTypeTxFeeBump(fee_bump_transaction) =>
```

Comment:
Should we also take `EnvelopeTypeTxFeeBump` into account?

Reply:
Hmm I don't think so. I doubt that we'll need that (soon or ever)